lib/handler/connect.c

```diff
@@ -765,8 +769,10 @@ static void udp_on_proceed(h2o_generator_t *_self, h2o_req_t *req)
         reset_io_timeout(self);
         h2o_socket_read_start(self->sock, udp_on_read);
     } else {
-        self->read_closed = 1;
-        h2o_send(self->src_req, NULL, 0, H2O_SEND_STATE_FINAL);
+        if (!self->read_closed) {
+            self->read_closed = 1;
+            h2o_send(self->src_req, NULL, 0, H2O_SEND_STATE_FINAL);
+        }
```

Comment:
Please see my discussion above.

I actually think that we might want to add `assert(!self->read_closed)` at the beginning of the function like we do in `tcp_on_proceed`.

Reply:
> I actually think that we might want to add `assert(!self->read_closed)` at the beginning of the function like we do in `tcp_on_proceed`.

I agree with your analysis that this is the only path that's actually lacking the guarantee that this PR claims exists. I'll close this and submit another PR.